# Battle of the Neighbourhoods - Toronto

The notebook contains all questions (1, 2 and 3) of the Assignment. The questions are separated by Section headers.

In [1]:
import pandas as pd

# Question 1

In [2]:
import requests

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

Response 200 means that we have a successful connection to the page.

In [4]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [5]:
len(wiki_data), type(wiki_data)

(3, list)

We need only the first table, so let's drop the other tables

In [6]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

We have Boroughs which are 'Not assigned', so let"s drop those rows.

In [7]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Let's group the records based on Postal Code.

In [8]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Check if we have records where Neighbourhood is 'Not assigned'.

In [9]:
df.Neighbourhood.str.count("Not assigned").sum()

0

0 means no 'Not assigned' records found, so we do not drop any rows. Let's reset index now.

In [10]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

We do not need 'index' column, let's drop it. 

In [11]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [12]:
df.shape

(103, 3)

**The answer for Question 1** is: we have 103 rows and 3 columns.

# Question 2

Let's install geocoder.

In [13]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.


Let's import it,

In [14]:
import geocoder

I have tried the below code for about 20 mins, then killed it. I have changed the cell to Markdown so that the run all execution doesn't stop.

```python
# initialize your variable to None
lat_lng_coords = None

postal_code = 'M3A'

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]
```

Alternatively, as suggested in the assignment, let's import the CSV file from the URL: https://cocl.us/Geospatial_data

In [15]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [16]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions of wiki data and csv data are the same, let's try to join on the postal codes to get the required data.

But before the join let's check the column types of both the dataframes, especially Postal Code column since we are trying to join on it.

In [17]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [18]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

Since Portal Code is object in both dataframes, let's try the join.

In [19]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [20]:
combined_data.shape

(103, 5)

**The answer for Question 2** is: we get 103 rows as expected when we do a inner join, so we have good data.

# Question 3

By inspiration from the previous lab where we cluster the neighbourhood of NYC, let's cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [21]:
from geopy.geocoders import Nominatim 

In [22]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Let's visualize the map of Toronto using folium.

In [23]:
import folium

In [24]:
# Creating the map
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Let's initialize Foursquare API credentials.

In [25]:
CLIENT_ID = 'CNMLUM1GGYHEK5ORQJUOJI3KKGUQ14EM2TCBLDJV24USVTDN' 
CLIENT_SECRET = 'UF20ZUIO3LOIK0IN51AM133DK2S04KEJATVMTGAMELHKAZFB'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: CNMLUM1GGYHEK5ORQJUOJI3KKGUQ14EM2TCBLDJV24USVTDN
CLIENT_SECRET:UF20ZUIO3LOIK0IN51AM133DK2S04KEJATVMTGAMELHKAZFB


Let's create a function to get all the venue categories in Toronto.

In [26]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Let's collect the venues in Toronto for each Neighbourhood.

In [27]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [28]:
venues_in_toronto.shape

(1330, 5)

So we have 1330 records and 5 columns. Let's check the sample data.

In [29]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2           GreenWin pool                   Pool  
3  Victoria Village Arena           Hockey Arena  
4               Portugril  Portuguese Restaurant

Let's check the Venues based on Neighbourhood.

In [30]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1316  Mimico NW, The Queensway West, South of Bloor,...   
1317  Mimico NW, The Queensway West, South of Bloor,...   
1318  Mimico NW, The Queensway West, South of Bloor,...   
1319  Mimico NW, The Queensway West, South of Bloor,...   
1320  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.753259               -79.329656           GreenWin pool   
3                  43.725882               -79.315572  Victoria Village Arena   
4                  43.725882               -79.315572               Portugril   
...                      ...                      ...                     ...   
1316               43.628841               -79.520999              Wingporium   
1317               43.628841               -79.520999        South St. Burger   
1318               43.628841               -79.520999               Dollarama   
1319               43.628841               -79.520999          Healthy Planet   
1320               43.628841               -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2                      Pool  
3              Hockey Arena  
4     Portuguese Restaurant  
...                     ...  
1316            Wings Joint  
1317           Burger Joint  
1318         Discount Store  
1319        Supplement Shop  
1320                 Bakery  

[415 rows x 5 columns]

We get 415 records for each neighbourhood. Let's check for the maximum venue categories.

In [31]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Adult Boutique                                   Church and Wellesley   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Adult Boutique                   43.665860               -79.383160   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.400049   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[237 rows x 4 columns]

Well, there are around 237 different types of Venue Categories. Good!

Let"s one-hot-encode the Venue Categories.

In [32]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
...                 ...             ...      ...                 ...   
1325                  0               0        0                   0   
1326                  0               0        0                   0   
1327                  0               0        0                   0   
1328                  0               0        0                   0   
1329                  0               0        0                   0   

      Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0                0               0                0                 0   
1                0               0                0                 0   
2                0               0                0                 0   
3                0               0                0                 0   
4                0               0                0                 0   
...            ...             ...              ...               ...   
1325             0               0                0                 0   
1326             0               0                0                 0   
1327             0               0                0                 0   
1328             0               0                0                 0   
1329             0               0                0                 0   

      American Restaurant  Antique Shop  ...  Train Station  \
0                       0             0  ...              0   
1                       0             0  ...              0   
2                       0             0  ...              0   
3                       0             0  ...              0   
4                       0             0  ...              0   
...                   ...           ...  ...            ...   
1325                    0             0  ...              0   
1326                    0             0  ...              0   
1327                    0             0  ...              0   
1328                    0             0  ...              0   
1329                    0             0  ...              0   

      Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                                 0                 0            0   
1                                 0                 0            0   
2                                 0                 0            0   
3                                 0                 0            0   
4                                 0                 0            0   
...                             ...               ...          ...   
1325                              0                 0            0   
1326                              0                 0            0   
1327                              0                 0            0   
1328                              0                 0            0   
1329                              0                 0            0   

      Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                         0                0         0            0   
1                         0                0         0            0   
2                         0                0         0            0   
3                         0                0         0            0   
4                         0                0         0            0   
...                     ...              ...       ...          ...   
1325                      0                0         0            0   
1326                      0                0         0            0   
1327                      0

Let's add the neighbourhood to the encoded dataframe.

In [33]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2         Parkwoods                  0               0        0   
3  Victoria Village                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Train Station  \
0                 0                    0  ...              0   
1                 0                    0  ...              0   
2                 0                    0  ...              0   
3                 0                    0  ...              0   
4                 0                    0  ...              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 238 columns]

Let's group the Neighbourhoods, let's calculate the mean venue categories in each Neighbourhood.

In [34]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0             0.0      0.0                 0.0           0.0             0.0   
1             0.0      0.0                 0.0           0.0             0.0   
2             0.0      0.0                 0.0           0.0             0.0   
3             0.0      0.0                 0.0           0.0             0.0   
4             0.0      0.0                 0.0           0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  ...  Train Station  \
0              0.0               0.0             0.000000  ...            0.0   
1              0.0               0.0             0.000000  ...            0.0   
2              0.0               0.0             0.000000  ...            0.0   
3              0.0               0.0             0.000000  ...            0.0   
4              0.0               0.0             0.041667  ...            0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 238 columns]

Let's make a function to get the top most common venue categories.

In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [36]:
import numpy as np

There are too many venue categories, let's take the top 10 to cluster the neighbourhoods.

In [37]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

       2nd Most Common Venue 3rd Most Common Venue    4th Most Common Venue  \
0  Latin American Restaurant          Skating Rink           Clothing Store   
1                        Gym        Sandwich Place             Dance Studio   
2                       Bank             Gift Shop       Chinese Restaurant   
3                       Bank    Chinese Restaurant      Japanese Restaurant   
4                Coffee Shop    Italian Restaurant  Comfort Food Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot    Dim Sum Restaurant  Ethiopian Restaurant   
1           Coffee Shop                   Pub              Pharmacy   
2           Bridal Shop        Sandwich Place                 Diner   
3           Yoga Studio  Ethiopian Restaurant           Escape Room   
4           Pizza Place                   Pub                  Café   

  8th Most Common Venue        9th Most Common Venue  \
0           Escape Room            Electronics Store   
1       Airport Service           College Rec Center   
2            Restaurant                Deli / Bodega   
3     Electronics Store  Eastern European Restaurant   
4      Sushi Restaurant            Indian Restaurant   

        10th Most Common Venue  
0  Eastern European Restaurant  
1         Ethiopian Restaurant  
2                  Supermarket  
3                    Drugstore  
4                   Restaurant

Let's make the model to cluster our Neighbourhoods. 

In [38]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [39]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

Let's check the labelling of our model.

In [40]:
kmeans.labels_[0:100]

array([4, 1, 1, 4, 1, 4, 4, 4, 4, 4, 0, 1, 4, 1, 4, 4, 4, 1, 1, 1, 4, 1,
       1, 4, 4, 4, 0, 1, 1, 1, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1, 4,
       4, 4, 4, 1, 4, 4, 0, 4, 4, 4, 4, 1, 4, 1, 4, 4, 4, 0, 1, 1, 1, 0,
       4, 4, 1, 4, 4, 4, 1, 4, 4, 4, 1, 1, 1, 4, 4, 2, 4, 1, 4, 1, 1, 3,
       4, 0, 4, 1, 1, 4, 0])

Let's add the clustering Label column to the top 10 common venue categories.

In [41]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Let's join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting.

In [42]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0                  Park   
1  43.725882 -79.315572             1.0           Coffee Shop   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             4.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                    Pool     Food & Drink Shop           Yoga Studio   
1   Portuguese Restaurant     French Restaurant          Hockey Arena   
2                    Park                Bakery        Breakfast Spot   
3  Furniture / Home Store     Accessories Store              Boutique   
4        Sushi Restaurant           Yoga Studio   Distribution Center   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0         Deli / Bodega     Electronics Store  Eastern European Restaurant   
1           Yoga Studio   Distribution Center                 Dessert Shop   
2     French Restaurant   Distribution Center                  Event Space   
3            Shoe Store    Miscellaneous Shop           Athletics & Sports   
4        Sandwich Place                  Park           Mexican Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0             Drugstore            Donut Shop                Dog Run  
1    Dim Sum Restaurant                 Diner         Discount Store  
2                   Spa        Farmers Market           Dessert Shop  
3          Carpet Store           Event Space            Coffee Shop  
4   Japanese Restaurant    Italian Restaurant             Hobby Shop

Let's drop all the NaN values to prevent data skew.

In [43]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Let's plot the clusters on the map.

In [44]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [45]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

**Cluster 1**

In [46]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             0.0                  Park   
21              York             0.0                  Park   
35         East York             0.0                  Park   
52        North York             0.0                  Park   
66        North York             0.0                  Park   
85       Scarborough             0.0                  Park   
91  Downtown Toronto             0.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Pool     Food & Drink Shop           Yoga Studio   
21         Women's Store                  Pool           Yoga Studio   
35     Convenience Store           Yoga Studio          Dessert Shop   
52           Yoga Studio      Department Store  Ethiopian Restaurant   
66     Convenience Store           Yoga Studio          Dessert Shop   
85            Playground          Intersection           Yoga Studio   
91                 Trail            Playground           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0          Deli / Bodega     Electronics Store  Eastern European Restaurant   
21      Department Store           Escape Room            Electronics Store   
35  Ethiopian Restaurant           Escape Room            Electronics Store   
52           Escape Room     Electronics Store  Eastern European Restaurant   
66  Ethiopian Restaurant           Escape Room            Electronics Store   
85      Department Store           Escape Room            Electronics Store   
91          Dessert Shop    Dim Sum Restaurant                        Diner   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                     Drugstore            Donut Shop                Dog Run  
21  Eastern European Restaurant             Drugstore             Donut Shop  
35  Eastern European Restaurant             Drugstore             Donut Shop  
52                    Drugstore            Donut Shop                Dog Run  
66  Eastern European Restaurant             Drugstore             Donut Shop  
85  Eastern European Restaurant             Drugstore             Donut Shop  
91               Discount Store   Distribution Center             Donut Shop

**Cluster 2**

In [47]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1         North York             1.0           Coffee Shop   
2   Downtown Toronto             1.0           Coffee Shop   
4   Downtown Toronto             1.0           Coffee Shop   
7         North York             1.0                   Gym   
13        North York             1.0                   Gym   
15  Downtown Toronto             1.0                  Café   
17         Etobicoke             1.0              Pharmacy   
22       Scarborough             1.0           Coffee Shop   
23         East York             1.0           Coffee Shop   
24  Downtown Toronto             1.0           Coffee Shop   
28        North York             1.0           Coffee Shop   
30  Downtown Toronto             1.0           Coffee Shop   
33        North York             1.0           Coffee Shop   
34        North York             1.0    Falafel Restaurant   
38       Scarborough             1.0      Department Store   
42  Downtown Toronto             1.0                  Café   
48  Downtown Toronto             1.0                  Café   
51       Scarborough             1.0                 Motel   
54      East Toronto             1.0           Coffee Shop   
55        North York             1.0        Sandwich Place   
56              York             1.0           Coffee Shop   
70         Etobicoke             1.0           Pizza Place   
72        North York             1.0              Pharmacy   
73   Central Toronto             1.0        Clothing Store   
74   Central Toronto             1.0                  Café   
76       Mississauga             1.0           Coffee Shop   
79   Central Toronto             1.0          Dessert Shop   
81      West Toronto             1.0                  Café   
86   Central Toronto             1.0           Coffee Shop   
93         Etobicoke             1.0           Pizza Place   
97  Downtown Toronto             1.0                  Café   

    2nd Most Common Venue 3rd Most Common Venue    4th Most Common Venue  \
1   Portuguese Restaurant     French Restaurant             Hockey Arena   
2                    Park                Bakery           Breakfast Spot   
4        Sushi Restaurant           Yoga Studio      Distribution Center   
7             Coffee Shop   Japanese Restaurant               Beer Store   
13            Coffee Shop   Japanese Restaurant               Beer Store   
15              Gastropub        Farmers Market              Coffee Shop   
17             Beer Store             Pet Store              Pizza Place   
22  Korean BBQ Restaurant           Yoga Studio             Dessert Shop   
23    Sporting Goods Shop          Burger Joint                     Bank   
24     Italian Restaurant                  Café                     Park   
28                   Bank             Gift Shop       Chinese Restaurant   
30                   Café           Pizza Place                Gastropub   
33         Clothing Store             Juice Bar               Restaurant   
34     Miscellaneous Shop  Caribbean Restaurant           Massage Studio   
38     Chinese Restaurant           Coffee Shop           Discount Store   
42            Coffee Shop            Restaurant      Japanese Restaurant   
48            Coffee Shop             Gastropub                    Hotel   
51    American Restaurant         Movie Theater            Deli / Bodega   
54                 Bakery                  Café      American Restaurant   
55            Coffee Shop    Italian Restaurant  Comfort Food Restaurant   
56         Discount Store        Sandwich Place               Restaurant   
70     Chinese Restaurant          Intersection              Coffee Shop   
72                Butcher           Pizza Place              Coffee Shop   
73            Coffee Shop            Bagel Shop     Fast Food Restaurant   
74         Sandwich Place           Coffee Shop               Donut Shop   
76                  Hotel                   Gym     

**Cluster 3**

In [48]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
98  Etobicoke             2.0                 River           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
98      Department Store  Ethiopian Restaurant           Escape Room   

   6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
98     Electronics Store  Eastern European Restaurant             Drugstore   

   9th Most Common Venue 10th Most Common Venue  
98            Donut Shop                Dog Run

**Cluster 4**

In [49]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
64             York             3.0         Jewelry Store   
68  Central Toronto             3.0                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
64           Yoga Studio        Farmers Market           Event Space   
68                  Park      Sushi Restaurant         Jewelry Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
64  Ethiopian Restaurant           Escape Room     Electronics Store   
68           Yoga Studio    Dim Sum Restaurant                 Diner   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
64  Eastern European Restaurant             Drugstore             Donut Shop  
68               Discount Store   Distribution Center             Donut Shop

**Cluster 5**

In [50]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
3          North York             4.0              Clothing Store   
6         Scarborough             4.0        Fast Food Restaurant   
8           East York             4.0                 Pizza Place   
9    Downtown Toronto             4.0                        Café   
10         North York             4.0         Japanese Restaurant   
12        Scarborough             4.0  Construction & Landscaping   
14          East York             4.0                Skating Rink   
16               York             4.0                       Trail   
18        Scarborough             4.0              Medical Center   
19       East Toronto             4.0            Asian Restaurant   
20   Downtown Toronto             4.0                 Coffee Shop   
25   Downtown Toronto             4.0               Grocery Store   
26        Scarborough             4.0          Athletics & Sports   
27         North York             4.0                 Golf Course   
29          East York             4.0              Sandwich Place   
31       West Toronto             4.0                      Bakery   
32        Scarborough             4.0                  Playground   
36   Downtown Toronto             4.0                        Park   
37       West Toronto             4.0                         Bar   
39         North York             4.0                        Café   
40         North York             4.0                        Park   
41       East Toronto             4.0            Greek Restaurant   
43       West Toronto             4.0                        Café   
44        Scarborough             4.0                      Bakery   
46         North York             4.0                        Park   
47       East Toronto             4.0              Sandwich Place   
49         North York             4.0                        Park   
50         North York             4.0                 Pizza Place   
53         North York             4.0                        Park   
57         North York             4.0              Baseball Field   
58        Scarborough             4.0             College Stadium   
59         North York             4.0            Ramen Restaurant   
60         North York             4.0                        Park   
61    Central Toronto             4.0                        Park   
62    Central Toronto             4.0                      Garden   
63               York             4.0                 Pizza Place   
65        Scarborough             4.0           Indian Restaurant   
67    Central Toronto             4.0                        Park   
69       West Toronto             4.0                        Park   
71        Scarborough             4.0   Middle Eastern Restaurant   
75       West Toronto             4.0              Breakfast Spot   
77          Etobicoke             4.0                        Park   
78        Scarborough             4.0                      Lounge   
80   Downtown Toronto             4.0                        Café   
82        Scarborough             4.0                    Pharmacy   
83    Central Toronto             4.0                Tennis Court   
84   Downtown Toronto             4.0                        Café   
87   Downtown Toronto             4.0             Airport Service   
88          Etobicoke             4.0                        Café   
89          Etobicoke             4.0               Grocery Store   
90        Scarborough             4.0        Fast Food Restaurant   
92   Downtown Toronto             4.0              Farmers Market   
94          Etobicoke             4.0                   Drugstore   
96   Downtown Toronto             4.0                 Coffee Shop   
99   Downtown Toronto             4.0                 Coffee Shop   
100      East Toronto             4.0                         Spa   
101         Etobicoke             4.0  Construction & Landscaping   
102         

We have **successfully** cluster Toronto neighbourhood based on venue categories!